### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by kuchma19**

### CODE REVIEW

File: test/Main.hs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import Test.Tasty (defaultMain, testGroup)
+import TestParser qualified
+import TestTemplate qualified
+
+main :: IO ()
+main =
+  defaultMain $
+    testGroup
+      "Lhx tests"
+      [ TestParser.tests,
```
Comment (astynax) [inline test/Main.hs]: We don't use trailing commas in this project.

File: test/TestTemplate.hs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        ],
+      testGroup
+        "right"
+        [ testCase "all good functions" $
+            isRight (L.makeTemplate "$rev:strip:lstrip:rstrip;")
+              @? "don't apply function in the list functions"
+        ]
+    ]
+
+createTestPropertyFunction :: (FunctionName, Char, T.Text -> T.Text) -> TestTree
```
Comment (astynax) [inline test/TestTemplate.hs]: In the library there are some functions named "makeSMTH", let's keep this style. Also, "createTest" looks strange. `makeFunctionPropertyTest`!

File: test/TestTemplate.hs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+createTestPropertyFunction (name, sep, f) = QS.testProperty (T.unpack name) $
+  \s -> not (T.any (== sep) s)
+    QS.==> fromMaybe False
+    $ rightToMaybe $ do
+      template <- L.makeTemplate ("$" <> name <> ";")
+      text <- L.apply template (L.makeInput (Separator $ T.singleton sep) s)
+      pure $ text == f s
+
+createTestPropertiesFunction :: [(FunctionName, Char, T.Text -> T.Text)] -> TestTree
+createTestPropertiesFunction functions =
```
Comment (astynax) [inline test/TestTemplate.hs]: I see exactly one usage of this function, let's "inline" it. Also, you have aliased `FunctionName` but what about `Char`. Its purpose has the same level of ambiguity, isn't it?

File: test/TestTemplate.hs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    ]
+
+applyTemplateWithIndexes :: TestTree
+applyTemplateWithIndexes =
+  testGroup
+    "indexes"
+    [ 
+      -- testCase "" $ assertEqual "$1 $0" (Right "b a") $ do 
+      --   template <- L.makeTemplate "$1 $0"
+      --   L.apply template (L.makeInput (Separator ",") "a,b")
```
Comment (astynax) [inline test/TestTemplate.hs]: A commented out code in PR, why?

File: test/TestParser.hs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import Test.Tasty.HUnit (testCase, (@?), (@?=))
+import Test.Tasty.QuickCheck as QC
+
+type Title = String
+
+type MessageIfFail = String
+
+type ParsedString = T.Text
+
+type ExpectedValue = [LP.Chunk]
```
Comment (astynax) [inline test/TestParser.hs]: This is a bad alias! In the library we already have one: `Template`. Why do we need another one just in tests? And why "Value"? All values we need to test are templates? I don't think so.

File: test/TestTemplate.hs (hunk 4 above)
Comment (kuchma19) [reply test/TestTemplate.hs]: I commented on the tests that failed

File: test/Main.hs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import TestTemplate qualified
+
+main :: IO ()
+main =
+  defaultMain $
+    testGroup
+      "Lhx tests"
+      [ TestParser.tests
+      , TestTemplate.tests
+      ]
```
Comment (astynax) [inline test/Main.hs]: Mandatory newline at the end of file is a good thing. Look at your editor's settings: it should have such option.

File: test/TestTemplate.hs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+applyFunctionTest =
+  let reverseTemplate = L.makeTemplate "$rev;"
+   in testGroup
+        "Apply tests"
+        [ QS.testProperty "reverse function" $
+            \s -> fromMaybe False $
+              rightToMaybe $ do
+                template <- reverseTemplate
+                text <- L.apply template (L.makeInput (Separator ",") s)
+                pure $ text == T.reverse s
```
Comment (astynax) [inline test/TestTemplate.hs]: It is a good taste to have all the properties declared at the top level. Or in `where` at least.

File: test/TestTemplate.hs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+tests :: TestTree
+tests =
+  testGroup
+    "Template tests"
+    [ makeTemplateTests
+    , makeInputTest
+    , applyFunctionTest
+    , applyTemplateWithIndexes
+    ]
```
Comment (astynax) [inline test/TestTemplate.hs]: No trailing newline again

File: test/TestTemplate.hs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                pure $ text == T.reverse s
+        ]
+
+applyTemplateWithIndexes :: TestTree
+applyTemplateWithIndexes =
+  let makeAssert templateT inputString result =
+        assertEqual (T.unpack templateT) (Right result) $ do
+          template <- L.makeTemplate templateT
+          L.apply template (L.makeInput (Separator ",") inputString)
+   in testGroup
```
Comment (astynax) [inline test/TestTemplate.hs]: I don't use `in` indented in such way :) And `where` will look nicer here: your helper function is not the first thing I'll want to see in the test body.

Also, "makeAssert" reads as "сделатьУтвердить" :) "makeAssertion" looks more literal. And usually one have such assertion functions named in a more declarative way — "shouldProduce" etc. But in this case I may recommend a nicer option:

```haskell
[ testCase "..." $
  "$0:rev;" `appliedTo` "abcd" @?= Right "dcba"
  ...
]
where
  appliedTo templateT inputString = ...
```

**Pull request merged**